### PR TITLE
[FIX] website(_*): better ondelete management on website_id

### PR DESCRIPTION
**Pull request by rdeodoo**
Before this commit, every model having website_id would have the default
`set null` as ondelete value for that field. Exception for `ir.ui.view` and
`website.redirect` which are set to `cascade`.

This meant that deleting a website would basically transform all its specific
records into generic ones.
That would lead to unwanted behavior, such as multiple `ir.attachment` with
same URL, when 2 websites had been theme-customized. Deleting one of the
website would result in later crashes when trying to open theme customization
in the remaining website(s), since those website would find their specific scss
custom attachment as well as the generic one from the deleted website.

Probably more behavior might be problematic when deleting a website before this
commit.

To summarize the choice of this commit implementation, every model not existing
outside website module are set to `cascade` (website.redirect, website.menu,
product.wishlist..).
The rest should be in restrict, as we don't want to delete a whole forum, blog
or job, which should probably be managed case by case to decide what to do.
Thus, restrict is a good choice to notify user of what should be handled.

Some exception remains, ir.ui.view should be delete on cascade while sale.order
should be set to null.

opw-2038414
opw-2035249

### CODE REVIEW

**Comment by rdeodoo** [comment]
As we discussed.

For the record, that's the list of `website_id` fields:

**Website_id coming from Mixin:**
- `blog.blog` -> add restrict
- `event.event` -> add restrict
- `forum.forum` -> add restrict
- `hr.job` -> add restrict
- `hr.employee` -> add restrict
- `delivery.carrier` -> add restrict
- `product.public.category` -> add restrict stable -> cascade master is possible to override in ORM
- `product.template` -> add restrict
- `slide.channel` -> add restrict

**Website_id hardcoded (no mixin)**
- **Related fields:**
  - `forum.post`: related to forum.forum
  - `stock.picking`: related to sale.order
  - `account.invoice`: related res.partner
  - `product.product`: related product.template
  - `slide.slide`: related slide.channel
  - `res.users` related: res.partner
- **Change default field attribute values:**
  - `product.wishlist`: required=True -> Add cascade
  - `product.pricelist`: default=.. -> Add restrict
  - `sale.order`: readonly/help -> keep set null
  - `sale.report`: readonly -> keep set null
  - `ir.ui.view`: ondelete=cascade -> Ok, already cascade
  - `website.redirect`: ondelete=cascade -> Ok, already cascade
- **No reason not to use mixin?:**
  - `payment.acquirer` -> add restrict
  - `website.twitter` -> add cascade
  - `ir.attachment` -> add cascade
  - `website.menu`: -> add cascade
  - `res.partner`: help -> add restrict

**Website_id coming from mixin but override**
- `blog.post`: related to blog.blog + multi.published.mixin
- `website.page`: related ir.ui.view + multi.published.mixin

Related fields are fine, as those will mimick what the related field does.
In master all non related field should use the mixin, only possible if we can override ondelete.

**Comment by rdeodoo** [comment]
My test sucks, if you have better idea..

File: addons/website/models/res_partner.py [hunk 1]
```diff
@@ -14,7 +14,7 @@ def urlplus(url, params):
 class Partner(models.Model):
     _inherit = "res.partner"
 
-    website_id = fields.Many2one('website', string='Registration Website')
+    website_id = fields.Many2one('website', string='Registration Website', ondelete='restrict')
```
Comment (JKE-be) [inline addons/website/models/res_partner.py]: setnull

File: addons/website/models/ir_attachment.py [hunk 2]
```diff
@@ -15,7 +15,7 @@ class Attachment(models.Model):
     # related for backward compatibility with saas-6
     website_url = fields.Char(string="Website URL", related='local_url', deprecated=True, readonly=False)
     key = fields.Char(help='Technical field used to resolve multiple attachments in a multi-website environment.')
-    website_id = fields.Many2one('website')
+    website_id = fields.Many2one('website', ondelete='cascade')
```
Comment (JKE-be) [inline addons/website/models/ir_attachment.py]: noo falback make it crappy

Need to override unlink:
key & website_Id = website
or .custom.... scss/js

File: addons/website/tests/test_ui.py [hunk 3]
```diff
@@ -5,6 +5,41 @@
 import odoo.tests
 
 
+@odoo.tests.tagged('-at_install', 'post_install')
+class TestUiCustomizeTheme(odoo.tests.HttpCase):
+    def test_01_attachment_website_unlink(self):
```
Comment (JKE-be) [inline addons/website/tests/test_ui.py]: attachment is now specific usecase,=>  + add test on other model than attchmnet

File: addons/website/models/res_partner.py (hunk 1 above)
Comment (rdeodoo) [reply addons/website/models/res_partner.py]: Done

File: addons/website/models/ir_attachment.py (hunk 2 above)
Comment (rdeodoo) [reply addons/website/models/ir_attachment.py]: Indeed, we don't want to unlink a SO pdf (that will only have its website_id flag if the website's domain is set and match the current URL ('bug' on get_current_website fallback))
Done

File: addons/website/tests/test_ui.py (hunk 3 above)
Comment (rdeodoo) [reply addons/website/tests/test_ui.py]: Test adapted, check that a PDF from a SO is not removed but only website_id empty, and a scss custo is removed.

File: addons/website/models/website.py [hunk 4]
```diff
@@ -154,6 +154,19 @@ def write(self, values):
             self.env['ir.qweb'].clear_caches()
         return result
 
+    @api.multi
+    def unlink(self):
+        # Do not delete invoices, delete what's strictly necessary
+        attachments_to_unlink = self.env['ir.attachment'].search([
```
Comment (rdeodoo) [inline addons/website/models/website.py]: Everyone agree with this domain?
Any theme attachment, theme customization or assets should be removed on website delete.
Anything else?

Comment (qsm-odoo) [reply addons/website/models/website.py]: Ok for me. As discussed, maybe we can make this more robust by sharing some of the code which adds this ".custom."... but maybe we should just take profit of my original idea of adding special tags on those attachment now that the document team had the right to do it :)

**Comment by JKE-be** [comment]
@qsm
If i’m not wrong these tag is in enterprise only ... no?

Jeremy Kersten

File: addons/website/models/website.py [hunk 5]
```diff
@@ -154,6 +154,19 @@ def write(self, values):
...
+            ('website_id', 'in', self.ids),
+            '|', '|',
+            ('key', '!=', False),  # theme attachment
+            ('url', 'ilike', '.custom.'),  # customized theme attachment
+            ('url', 'ilike', '.assets_'),
```
Comment (JKE-be) [inline addons/website/models/website.py]: .assets\_

```python
In [0]: x = 'website.assetssansunderscore'                                                                                                                                                                        

In [1]: env['ir.attachment'].create({'name': x, 'url':x})                                                                                                                                                         
Out[1]: ir.attachment(538,)

In [2]: env['ir.attachment'].search([('url', 'ilike', '.assets_')], order='id desc', limit=1).mapped('url')                                                                                                       
Out[2]: ['website.assetssansunderscore']

In [3]: env['ir.attachment'].search([('url', 'ilike', '.assets\_')], order='id desc', limit=1).mapped('url')                                                                                                      
Out[3]: ['/web/content/535-9a90afe/website.assets_editor.js']

```

+ indent the '|'

Comment (rdeodoo) [reply addons/website/models/website.py]: Good catch, indeed it was not doing exactly what I wanted.
Note that I wrote it `.assets\\_` for linter escaping error (not sure there is a difference aside linter error between one or two `\` in python string, while on other cases their might be [1](https://stackoverflow.com/questions/38084864/how-to-escape-underscores-in-postgresql#comment70198430_38084865) [2](https://github.com/mysqljs/mysql/issues/650) ).

About the indent of `|`, do you prefer it written in 2 lines instead of one? Not sure if that's what u meant.

File: addons/website/models/website.py [hunk 6]
```diff
@@ -154,6 +154,19 @@ def write(self, values):
...
 
+    @api.multi
+    def unlink(self):
+        # Do not delete invoices, delete what's strictly necessary
+        attachments_to_unlink = self.env['ir.attachment'].search([
+            ('website_id', 'in', self.ids),
+            '|', '|',
+            ('key', '!=', False),  # theme attachment
+            ('url', 'ilike', '.custom.'),  # customized theme attachment
+            ('url', 'ilike', '.assets\\_'),
```
Comment (JKE-be) [inline addons/website/models/website.py]: strange syntax, but seems to works...

Comment (JKE-be) [reply addons/website/models/website.py]:
```
('website_id', 'in', self.ids),
'|', '|',
       ('key', '!=', False),  # theme attachment
       ('url', 'ilike', '.custom.'),  # customized theme attachment
       ('url', 'ilike', '.assets\\_'),
])
```

**Comment by JKE-be** [comment]
@robodoo r+

Pour le truc de rendre callable les attributes comme ondelete, rco trouve ca con et que c'est de la fainéantise de ne pas vouloir dupliquer les champs et donc dupliquer les label and co pour rien...
Donc ca reste ainsi pour master aussi.

La mixin c'est fait pour centraliser du code, et pas des champs...

**Comment by rdeodoo** [comment]
Mouais.. Comme ça la moitié des modèles ont website_id sans la mixin et donc tu penses pouvoir call les méthodes de la mixin ou pouvoir te baser sur les behavior de celle ci comme published mais en fait non..

La moitié des trucs introduits dans les nouveaux devs website des autres équipes n'ont jamais la mixin.

Fin bon on peut pas mener toutes les batailles.

Ty for merge

**Comment by qsm-odoo** [comment]
:sob:

**Comment by JKE-be** [comment]
Faut juste attendre que qqun d'autre en aie besoin, et a ce moment peut etre on trouvera que c'est une bonne idée... 
:disappointed:
